### PR TITLE
Remove deprecated class MGDoFHandler

### PR DESCRIPTION
**Pull request by bangerth**
This actually took a good bit more work than I thought it would. Principally, there are a lot of places where the semantics of functions that were implemented in MGDoFHandler were different from the corresponding functions in DoFHandler. This requires going through a lot of testcases and adjusting them.

### CODE REVIEW

File: source/multigrid/CMakeLists.txt [hunk 1]
```diff
@@ -1,6 +1,6 @@
 ## ---------------------------------------------------------------------
 ##
-## Copyright (C) 2012 - 2013 by the deal.II authors
+## Copyright (C) 2012 - 2013, 2015 by the deal.II authors
```
Comment (tamiko) [inline source/multigrid/CMakeLists.txt]: Can we just stick to <code>Copyright (C) 2012 - 2015</code>?

Comment (bangerth) [inline source/multigrid/CMakeLists.txt]: Would you allow me to do this as a follow-up? In fact, I think we should always tag every file's copyright notice for every release.

File: tests/benchmarks/old/patches/library/2007-05-30 [hunk 2]
```diff
@@ -27,14 +27,14 @@ Index: deal.II/source/multigrid/mg_dof_handler.cc
...
  }
  
  
 +template <int dim>
 +unsigned int
-+MGDoFHandler<dim>::memory_consumption() const
++DoFHandler<dim>::memory_consumption() const
 +{
 +  unsigned int mem = DoFHandler<dim>::memory_consumption();
 +  for (unsigned int l=0;l<mg_levels.size();++l)
```
Comment (tamiko) [inline tests/benchmarks/old/patches/library/2007-05-30]: *hugh* What about removing this patch file altogether? Or keeping it around unchanged for historical/sentimental reasons? It surely doesn't apply anymore (given the fact that mg_dof_handler.cc is removed by this very pull request).

File: tests/bits/step-16.cc [hunk 3]
```diff
@@ -1,6 +1,6 @@
 // ---------------------------------------------------------------------
 //
-// Copyright (C) 2005 - 2013 by the deal.II authors
+// Copyright (C) 2005 - 2013, 2015 by the deal.II authors
```
Comment (tamiko) [inline tests/bits/step-16.cc]: same here with the copyright line.

I thought we have agreed on keeping it as simple as possible, i.e. first [year of introduction] - [last year of modification]. If not, I have already changed around 1000 file headers :-]

File: tests/deal.II/block_info.cc [hunk 4]
```diff
@@ -1,6 +1,6 @@
 // ---------------------------------------------------------------------
 //
-// Copyright (C) 2009 - 2013 by the deal.II authors
+// Copyright (C) 2009 - 2013, 2015 by the deal.II authors
```
Comment (tamiko) [inline tests/deal.II/block_info.cc]: as above

File: tests/benchmarks/old/patches/library/2007-05-30 (hunk 2 above)
Comment (bangerth) [inline tests/benchmarks/old/patches/library/2007-05-30]: Yes, I think the files in this directory can go altogether. Let's leave it for a separate patch.

**Comment by tamiko** [comment]
What a ton of work! Looks fine.
I'd suggest to remove this old benchmark patch, or just leave it as it is.

Please rebase onto master and merge.

File: tests/bits/step-16.cc (hunk 3 above)
Comment (bangerth) [inline tests/bits/step-16.cc]: The problem is that most of our patches don't update the copyright year. emacs does it for me, but I only sometimes use emacs. Most of my patches are written in eclipse, and eclipse doesn't update the copyright year. So I think that for most files, the current state does not adequately reflect the year anyway. Thus my suggestion to do it ahead of every release for all files.

**Comment by bangerth** [comment]
Why do you suggest to rebase? I can merge through the web interface, so a FF merge seems possible, no?

**Comment by tamiko** [comment]
> Why do you suggest to rebase? I can merge through the web interface,
> so a FF merge seems possible, no?

Just for cosmetic reasons. You're more than 30 commits behind.

But feel free to just merge without rebasing.

Best,
Matthias

**Comment by bangerth** [comment]
OK, I'll update and merge.

May git-fu may be limited, but I was under the impression that all web-based github merges are fast-forward. So how would it look different between rebasing and not?

**Comment by tjhei** [comment]
> I was under the impression that all web-based github merges are fast-forward

You can only fast forward if your commits go back to the _current_ HEAD, so no. Github will allow merging if now conflicts appear when doing ```git merge```. If you don't rebase the history will look a bit more ugly.

File: doc/news/changes.h [hunk 5]
```diff
@@ -140,7 +140,30 @@ inconvenience this causes.
...
+  functionality of this class had previously been incorporated into
+  the DoFHandler class. Unlike the changes above, if you were still
+  using this class, you will need to do the following changes to
+  your code:
+  - Where you called <code>mg_dof_handler.distribute_dofs()</code>
+    you now also need to explicitly call
+    <code>mg_dof_handler.distribute_mg_dofs()</code>.
+  - If you called <code>mg_dof_handler.begin(level)</code>, you
+    will now have to write this as
+    <code>mg_dof_handler.begin_mg(level)</code> to make clear that
```
Comment (tjhei) [inline doc/news/changes.h]: Is this really the case? I think you can still use ```begin(level)```. The difference is that you get a ```level_cell_iterator``` with ```begin_mg```, which makes changes the semantic of ```get_dof_indices``` be what the old ```get_mg_dof_indices``` did.

Comment (bangerth) [inline doc/news/changes.h]: The issue is in places where you call something like
```
  MeshWorker::loop (mg_dof_handler.begin(level), mg_dof_handler.end(level), ...);
```
Before, the argument types where level iterators (because the local typedef of `MGDoFHandler` was a level iterator) and MeshWorker could successfully assemble the level matrices. But now we use the `DoFHandler` typedef, which is not a level iterator, and so MeshWorker will call operations that are not valid on non-level iterators. So in contexts where the argument type is automatically deduced, you really need to do this. 

Of course, in places where you write
```
  DoFHandler::level_cell_iterator cell = dh.begin(level);
```
everything is fine because the iterator will be converted as appropriate.

Comment (tjhei) [inline doc/news/changes.h]: Okay, I forgot that we already adjusted MeshWorker to take advantage of level iterators. Thanks for clarifying.
